### PR TITLE
MARC holdings - Leader rules

### PR DESCRIPTION
**Pull request by Serhii-Zdorenko**
### Purpose
Marc holdings leader rules

### Approach
create validator for leader

### Learning
https://issues.folio.org/browse/MODQM-151

### CODE REVIEW

File: src/main/java/org/folio/qm/converter/elements/LeaderItem.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  ENTRY_MAP_21("Entry map", 21, 1, '5'),
+  ENTRY_MAP_22("Entry map", 22, 1, '0'),
+  ENTRY_MAP_23("Entry map", 23, 1, '0'),
+
+  //COMMON FIELDS FOR HOLDINGS AND AUTHORITY
+  UNDEFINED_CHARACTER_POSITION_7("Undefined character position", 7, 1, '\\', ' ', '\u00A0'),
+  UNDEFINED_CHARACTER_POSITION_8("Undefined character position", 8, 1, '\\', ' ', '\u00A0'),
+  UNDEFINED_CHARACTER_POSITION_19("Undefined character position", 19, 1, '\\', ' ', '\u00A0'),
+
+  //BIB FIELDS
```
Comment (oleksandr-dekin) [inline src/main/java/org/folio/qm/converter/elements/LeaderItem.java:24]: Maybe will be better to separate enum for holding , bib and authority?

File: NEWS.md [hunk 2]
```diff
@@ -8,6 +8,7 @@
 * [MODQM-146](https://issues.folio.org/browse/MODQM-146) - View MARC Holdings via quickMARC
 * [MODQM-142](https://issues.folio.org/browse/MODQM-142) - Backend: Edit MARC Holdings via quickMARC
 * [MODQM-150](https://issues.folio.org/browse/MODQM-150) - MARC holdings - 008 rules
+* [MODQM-151](https://issues.folio.org/browse/MODQM-151) - MARC holdings - Leader rules
```
[the remark this answers is not in the document]
Comment (Serhii-Zdorenko) [reply NEWS.md:11]: No, I close that task without PR.

File: src/main/java/org/folio/qm/validation/LeaderValidationRule.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    if (error.isPresent()) {
+      return error;
+    }
+    error = validateLeaderNumberFields(leader, BASE_ADDRESS.getPosition(), BASE_ADDRESS.getLength());
+    if (error.isPresent()) {
+      return error;
+    }
+    return validateLeaderFieldsRestrictions(leader, List.of(CODING_SCHEME, INDICATOR_COUNT, SUBFIELD_CODE_LENGTH,
+      ENTRY_MAP_20, ENTRY_MAP_21, ENTRY_MAP_22, ENTRY_MAP_23));
+  }
```
Comment (natali-zaitseva) [inline src/main/java/org/folio/qm/validation/LeaderValidationRule.java:42]: could you change it to 
return  Stream.of(
        validateLeaderLength(leader),
        validateLeaderNumberFields(leader, RECORD_LENGTH.getPosition(), RECORD_LENGTH.getLength()),
        validateLeaderNumberFields(leader, BASE_ADDRESS.getPosition(), BASE_ADDRESS.getLength()),
        validateLeaderFieldsRestrictions(leader, List.of(CODING_SCHEME, INDICATOR_COUNT, SUBFIELD_CODE_LENGTH,
            ENTRY_MAP_20, ENTRY_MAP_21, ENTRY_MAP_22, ENTRY_MAP_23)))
        .filter(Optional::isPresent)
        .map(Optional::get)
        .findFirst();

File: src/main/java/org/folio/qm/validation/LeaderValidationRule.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+  private Optional<ValidationError> validateLeaderLength(String leader) {
+    return leader.length() == Constants.LEADER_LENGTH ? Optional.empty() : Optional.of(createValidationError(leader, "Wrong leader length"));
```
Comment (natali-zaitseva) [inline src/main/java/org/folio/qm/validation/LeaderValidationRule.java:45]: what happens if the leader is null?

Comment (Serhii-Zdorenko) [reply src/main/java/org/folio/qm/validation/LeaderValidationRule.java]: good point, corrected

**Review by natali-zaitseva — APPROVED** [review]

**Pull request merged**